diff --git a/lib/irc.js b/lib/irc.js
index 5abc7f7..8d287a7 100644
--- a/lib/irc.js
+++ b/lib/irc.js
@@ -499,12 +499,6 @@ function Client(server, nick, opt) {
                 }
                 break;
             case 'PART':
-                // channel, who, reason
-                self.emit('part', message.args[0], message.nick, message.args[1], message);
-                self.emit('part' + message.args[0], message.nick, message.args[1], message);
-                if (message.args[0] != message.args[0].toLowerCase()) {
-                    self.emit('part' + message.args[0].toLowerCase(), message.nick, message.args[1], message);
-                }
                 if (self.nick == message.nick) {
                     channel = self.chanData(message.args[0]);
                     delete self.chans[channel.key];
@@ -515,16 +509,16 @@ function Client(server, nick, opt) {
                         delete channel.users[message.nick];
                     }
                 }
-                break;
-            case 'KICK':
-                // channel, who, by, reason
-                self.emit('kick', message.args[0], message.args[1], message.nick, message.args[2], message);
-                self.emit('kick' + message.args[0], message.args[1], message.nick, message.args[2], message);
+
+                // channel, who, reason
+                self.emit('part', message.args[0], message.nick, message.args[1], message);
+                self.emit('part' + message.args[0], message.nick, message.args[1], message);
                 if (message.args[0] != message.args[0].toLowerCase()) {
-                    self.emit('kick' + message.args[0].toLowerCase(),
-                              message.args[1], message.nick, message.args[2], message);
+                    self.emit('part' + message.args[0].toLowerCase(), message.nick, message.args[1], message);
                 }

+                break;
+            case 'KICK':
                 if (self.nick == message.args[1]) {
                     channel = self.chanData(message.args[0]);
                     delete self.chans[channel.key];
@@ -535,6 +529,15 @@ function Client(server, nick, opt) {
                         delete channel.users[message.args[1]];
                     }
                 }
+
+                // channel, who, by, reason
+                self.emit('kick', message.args[0], message.args[1], message.nick, message.args[2], message);
+                self.emit('kick' + message.args[0], message.args[1], message.nick, message.args[2], message);
+                if (message.args[0] != message.args[0].toLowerCase()) {
+                    self.emit('kick' + message.args[0].toLowerCase(),
+                              message.args[1], message.nick, message.args[2], message);
+                }
+
                 break;
             case 'KILL':
                 nick = message.args[0];
